File: ascii_gm/text_generator.py
import random
import re

# based on https://donjon.bin.sh/code/random/


def generate_text(input_type, gen_data):
    list = gen_data.get(input_type)
    if list:
        string = select_from(list)
        if string:
            return expand_tokens(string, gen_data)
    return ""


def generate_list(input_type, n_of):
    list = []
    for i in range(n_of):
        list.append(generate_text(input_type))
    return list


def select_from(input_list):
    if isinstance(input_list, list):
        return select_from_array(input_list)
    else:
        return select_from_table(input_list)


def select_from_array(list):
    return list[random.randint(0, len(list) - 1)]


def select_from_table(list):
    len = scale_table(list)
    if len:
        idx = random.randint(1, len)
        for key in list:
            r = key_range(key)
            if idx >= r[0] and idx <= r[1]:
                return list[key]
    return ""


def scale_table(list):
    len = 0
    for key in list:
        r = key_range(key)
        len = max(len, r[1])
    return len


def key_range(key):
    match = re.match(r"(\d+)-00", key)
    if match:
        return [int(match.group(1)), 100]
    match = re.match(r"(\d+)-(\d+)", key)
    if match:
        return [int(match.group(1)), int(match.group(2))]
    if key == "00":
        return [100, 100]
    return [int(key), int(key)]


def expand_tokens(string, gen_data):
    match = re.search(r"{(\w+)}", string)
    while match:
        token = match.group(1)
        repl = generate_text(token, gen_data)
        if repl:
            string = string.replace("{" + token + "}", repl)
        else:
            string = string.replace("{" + token + "}", token)
        match = re.search(r"{(\w+)}", string)
    return string

File: lib/text_generator.ts
import { random, randomInt } from "./rng.ts";

export function keyRange(key: string): [number, number] {
  const m00 = key.match(/^(\d+)-00$/);
  if (m00) return [parseInt(m00[1]), 100];
  const mr = key.match(/^(\d+)-(\d+)$/);
  if (mr) return [parseInt(mr[1]), parseInt(mr[2])];
  if (key === "00") return [100, 100];
  return [parseInt(key), parseInt(key)];
}

function scaleTable(table: Record<string, unknown>): number {
  let max = 0;
  for (const key of Object.keys(table)) {
    const r = keyRange(key);
    max = Math.max(max, r[1]);
  }
  return max;
}

function selectFromArray<T>(arr: T[]): T {
  return arr[Math.floor(random() * arr.length)];
}

function selectFromTable(table: Record<string, unknown>): unknown {
  const len = scaleTable(table);
  if (!len) return "";
  const idx = Math.floor(random() * len) + 1;
  for (const key of Object.keys(table)) {
    const r = keyRange(key);
    if (idx >= r[0] && idx <= r[1]) return table[key];
  }
  return "";
}

function selectFrom(input: unknown): unknown {
  if (Array.isArray(input)) return selectFromArray(input);
  return selectFromTable(input as Record<string, unknown>);
}

function expandTokens(string: string, genData: Record<string, unknown>): string {
  let result = string;
  let match = result.match(/{(\w+)}/);
  while (match) {
    const token = match[1];
    const repl = generateText(token, genData);
    result = result.replace(`{${token}}`, repl ?? token);
    match = result.match(/{(\w+)}/);
  }
  return result;
}

export function generateText(inputType: string, genData: Record<string, unknown>): string | null {
  const list = genData[inputType];
  if (list) {
    const string = selectFrom(list);
    if (string) return expandTokens(string as string, genData);
  }
  return null;
}

export function generateList(inputType: string, n: number, genData: Record<string, unknown>): (string | null)[] {
  const results: (string | null)[] = [];
  for (let i = 0; i < n; i++) {
    results.push(generateText(inputType, genData));
  }
  return results;
}
